src/training/communicator.h

```diff
@@ -175,7 +175,17 @@ class DefaultCommunicator : public ICommunicator {
       }
     };
 
+    // reset gradients outside current shard
+    auto reset = [this, shardSize](size_t idx, size_t begin, size_t end) {
+      auto grad = graphs_[idx]->params()->grads();
+      if (begin > 0)
+        grad->subtensor(0, begin)->set(0);
+      if (end < grad->size())
+        grad->subtensor(end, grad->size()-end)->set(0);
+    };
+
     foreach(scatter);
+    foreach(reset);
```

Comment:
Where does this change come from?

Reply:
By cherry-picking communicator.h from master in the latest merge with master. The change comes from 
https://github.com/marian-nmt/marian-dev/commit/2111c280362fcf1b76bb13688e5a8f372e435479

Reply:
I actually don't understand why github lists this as a difference.

Reply:
I have been running in to this recently. Seems merge diffs are done against last common ancestor or something. Makes no sense to me either. At least give me a way to see an actual 3-way diff, not a fictional one.